zetaclient/chains/evm/observer/v2_inbound.go

```diff
@@ -149,7 +151,58 @@ func (ob *Observer) parseAndValidateDepositEvents(
 		return validEvents[i].Raw.BlockNumber < validEvents[j].Raw.BlockNumber
 	})
 
-	return validEvents
+	// Check if multiple calls are enabled
+	if ob.shouldAllowMultipleCalls(ctx) {
+		return validEvents
+	}
+	return ob.filterDepositEventsByTx(validEvents)
+}
+
+// shouldAllowMultipleCalls checks the EnableMultipleCalls feature flag
+func (ob *Observer) shouldAllowMultipleCalls(ctx context.Context) bool {
+	// Get app context to access feature flags
+	app, err := zctx.FromContext(ctx)
+	if err != nil {
+		ob.Logger().Inbound.Warn().Err(err).Msg("unable to get app context, using default behavior")
+		// Fallback to filtering (return false)
+		return false
+	}
+
+	// Check feature flag to determine whether to allow multiple calls from same tx
+	cfg := app.Config()
+
+	// If EnableMultipleCalls is enabled, return all valid events without filtering
+	if cfg.IsEnableMultipleCallsEnabled() {
+		ob.Logger().Inbound.Debug().
+			Msg("EnableMultipleCalls enabled: returning all valid events without filtering")
+		return true
+	}
+
+	// Default behavior: filter events from same tx (only first event)
+	ob.Logger().Inbound.Debug().
+		Msg("EnableMultipleCalls disabled: filtering multiple events from same tx (only first event)")
+	return false
+}
+
+// filterDepositEventsByTx filters deposit events from same tx (original logic)
+func (ob *Observer) filterDepositEventsByTx(
```

Comment:
The name is a bit unclear to me, imo we don't really filter the even, we should purposefully block subsequent event in a deposit.

Applying the guard is the only logic for filtering I would use a name that reflect this like `ApplyMultiDepositGuard`

Reply:
i refactored how ff are fetched, and reverted this code back into function like it was before multiple calls